src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs

```diff
@@ -334,36 +335,45 @@ public override IReadOnlyList<TextChangeRange> GetChangeRanges(SourceText oldTex
                     }
                 }
 
-                var oldSnapshot = oldText.FindCorrespondingEditorTextSnapshot();
-                var newSnapshot = this.FindCorrespondingEditorTextSnapshot();
+                var oldSnapshot = oldText.TryFindCorrespondingEditorTextImage();
+                var newSnapshot = this.TryFindCorrespondingEditorTextImage();
                 return GetChangeRanges(oldSnapshot, oldText.Length, newSnapshot);
             }
 
-            private IReadOnlyList<TextChangeRange> GetChangeRanges(ITextSnapshot oldSnapshot, int oldTextLength, ITextSnapshot newSnapshot)
+            private IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImage oldImage, int oldTextLength, ITextImage newImage)
             {
-                if (oldSnapshot == null ||
-                    newSnapshot == null ||
-                    oldSnapshot.TextBuffer != newSnapshot.TextBuffer)
+                var oldSnapshot = oldImage == null ? null : TextBufferMapper.TryFindEditorSnapshot(oldImage);
+                var newSnapshot = newImage == null ? null : TextBufferMapper.TryFindEditorSnapshot(newImage);
+
+                if (oldImage == null ||
+                    newImage == null ||
+                    AreFromDifferentBuffer(oldSnapshot, newSnapshot))
                 {
                     // Claim its all changed
                     Logger.Log(FunctionId.Workspace_SourceText_GetChangeRanges, "Invalid Snapshots");
-                    return ImmutableArray.Create<TextChangeRange>(new TextChangeRange(new TextSpan(0, oldTextLength), this.Length));
+                    return ImmutableArray.Create(new TextChangeRange(new TextSpan(0, oldTextLength), this.Length));
                 }
-                else if (oldSnapshot.Version.ReiteratedVersionNumber == newSnapshot.Version.ReiteratedVersionNumber)
+                else if (AreSameReiteratedVersion(oldSnapshot, newSnapshot))
                 {
                     // content of two snapshot must be same even if versions are different
                     return TextChangeRange.NoChanges;
                 }
                 else
                 {
-                    return GetChangeRanges(oldSnapshot, newSnapshot, forward: oldSnapshot.Version.VersionNumber <= newSnapshot.Version.VersionNumber);
+                    return GetChangeRanges(oldImage, newImage, forward: oldImage.Version.VersionNumber <= newImage.Version.VersionNumber);
                 }
             }
 
+            private static bool AreSameReiteratedVersion(ITextSnapshot oldSnapshot, ITextSnapshot newSnapshot) 
```

Comment:
due to reference equality, I think we should remove this optimization. data might be same but two are 2 different instances which will mess up our map to get back to editor buffer/snapshot.